### PR TITLE
Add support for `ERC20` contracts to Gateway

### LINKED ISSUE #76: Gateway `ERC20` support
Gateway should be able to support the `ERC6160` standard as well as the `ERC20` standard through token swaps. If a user bridges an `ERC20` token that has no `ERC20` counterpart on the destination chain, they should get an `ERC6160` token. On the other hand if a counterpart exists, the relayers should double as liquidity providers and facilitate the token swap. Relayers after providing liquidity should get the `ERC6160` counterpart that they can use to withdraw the `ERC20` token locked on the source chain.

`onRequestTimeout` we should either refund the custodied `ERC20` or mint the `ERC6160` asset back to the user. Allowing them retry the cross-chain teleportation.

Because ISMP charges protocol fees in it's configured `feeToken`, Gateway should allow users swap either the native asset or an `ERC20` for the `feeToken` asset through some local AMM required to cover their fees.

### PR DESCRIPTION
**Pull request by seunlanlege**
Closes #76, #77

### CODE REVIEW

File: evm/script/DeployIsmp.s.sol [hunk 1]
```diff
@@ -81,7 +81,5 @@ contract DeployScript is Script {
             PolygonHost h = new PolygonHost{salt: salt}(params);
             return address(h);
         }
-
-        revert("unknown host");
```
Comment (seunlanlege) [inline evm/script/DeployIsmp.s.sol:85]: Why remove?

File: evm/src/modules/HostManager.sol [hunk 2]
```diff
@@ -40,17 +45,19 @@ contract HostManager is IIsmpModule {
...
+        // note, the below will revert with solidity error `Panic(0x21)`
+        OnAcceptActions action = OnAcceptActions(uint8(request.body[0]));
+        if (action == OnAcceptActions.Withdraw) {
             // This is where relayers can withdraw their fees.
-            WithdrawParams memory params = abi.decode(request.body, (WithdrawParams));
+            WithdrawParams memory params = abi.decode(request.body[1:], (WithdrawParams));
             IHostManager(_params.host).withdraw(params);
         } else {
-            HostParams memory params = abi.decode(request.body, (HostParams));
+            // i.e if (action == OnAcceptActions.SetHostParam)
```
Comment (seunlanlege) [inline evm/src/modules/HostManager.sol:59]: prefer if this isn't implicitly assumed and we indeed check the message type

File: evm/src/modules/TokenGateway.sol [hunk 3]
```diff
@@ -48,51 +77,109 @@ contract TokenGateway is IIsmpModule {
     // set the ismp host address
     function setIsmpHost(address _host) public onlyAdmin {
         host = _host;
-        admin = address(0);
+    }
+
+    function setTokenIdentifierDetails(bytes32 _tokenId, address _erc20, address _erc6160) external onlyAdmin {
+        _erc6160s[_tokenId] = _erc6160;
+        _erc20s[_tokenId] = _erc20;
+    }
```
Comment (seunlanlege) [inline evm/src/modules/TokenGateway.sol:85]: Should be done individually

File: evm/src/GovernanceToken.sol [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return true;
+    }
+
+    function onAccept(PostRequest calldata request) external onlyRole(HOST_ROLE) {
+        (address account, bytes32 role, bool grant) = abi.decode(request.body, (address, bytes32, bool));
+        grant ? _grantRole(role, account) : _revokeRole(role, account);
+    }
+
+    function onPostResponse(PostResponse memory) external view onlyRole(HOST_ROLE) {
+        revert("Token gateway doesn't emit Post responses");
```
Comment (seunlanlege) [inline evm/src/GovernanceToken.sol:34]: Token gateway?

File: evm/src/modules/HostManager.sol (hunk 2 above)
Comment (seunlanlege) [reply evm/src/modules/HostManager.sol:59]: unresolved

File: evm/src/modules/TokenGateway.sol (hunk 3 above)
Comment (seunlanlege) [reply evm/src/modules/TokenGateway.sol:93]: unresolved

File: evm/src/modules/TokenGateway.sol [hunk 5]
```diff
@@ -41,58 +77,137 @@ contract TokenGateway is IIsmpModule {
...
+                unchecked {
+                    toBridge -= params.amountForFee;
+                }
+            }
+        } else if (erc6160 != address(0) && params.redeem) {
+            // we're sending an erc6160 asset so we should redeem on the destination if we can.
+            IERC6160Ext20(erc6160).burn(from, params.amount, "");
+        } else {
+            revert("Gateway: Unknown Token Identifier");
+        }
```
Comment (seunlanlege) [inline evm/src/modules/TokenGateway.sol:139]: protocol fees?

File: evm/src/GovernableToken.sol [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        _roles[role][account] = false;
+    }
+
+    function mint(address account, uint256 amount, bytes calldata) external override onlyRole(MINTER_ROLE) {
+        _mint(account, amount);
+    }
+
+    function burn(address account, uint256 amount, bytes calldata) external override onlyRole(BURNER_ROLE)  {
+        _burn(account, amount);
+    }
```
Comment (seunlanlege) [inline evm/src/GovernableToken.sol:33]: The ERC6160 interface already provides these methods

File: evm/src/GovernableToken.sol [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    function onAccept(PostRequest calldata request) external onlyRole(HOST_ROLE) {
```
Comment (seunlanlege) [inline evm/src/GovernableToken.sol:35]: uhm, the host is a contract

File: evm/src/GovernableToken.sol [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        (address account, bytes32 role, bool grant) = abi.decode(request.body, (address, bytes32, bool));
+        grant ? _grantRole(role, account) : _revokeRole(role, account);
+    }
+
+    function onPostResponse(PostResponse memory) external view onlyRole(HOST_ROLE) {
+        revert("Token gateway doesn't emit Post responses");
```
Comment (seunlanlege) [inline evm/src/GovernableToken.sol:41]: Token gateway?

File: evm/src/modules/TokenGateway.sol [hunk 9]
```diff
@@ -41,58 +77,137 @@ contract TokenGateway is IIsmpModule {
...
+        bytes32 localAsset = _assets[body.tokenId];
+        address erc20 = _erc20s[localAsset];
+        address erc6160 = _erc6160s[localAsset];
+
+        // prefer to give the user erc20
+        if (erc20 != address(0) && body.redeem) {
+            // a relayer/user is redeeming the native asset
+            require(IERC20(erc20).transfer(body.to, body.amount), "Gateway: Insufficient Balance");
+        } else if (erc20 != address(0) && erc6160 != address(0)) {
+            // relayers double as liquidity providers, todo: protocol fees
```
Comment (seunlanlege) [inline evm/src/modules/TokenGateway.sol:171]: protocol fees?

File: evm/src/GovernableToken.sol [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    function mint(address account, uint256 amount, bytes calldata) external override onlyRole(MINTER_ROLE) {
+        _mint(account, amount);
+    }
+
+    function burn(address account, uint256 amount, bytes calldata) external override onlyRole(BURNER_ROLE)  {
+        _burn(account, amount);
+    }
+
+    function onAccept(PostRequest calldata request) external onlyRole(HOST_ROLE) {
+        (address account, bytes32 role, bool grant) = abi.decode(request.body, (address, bytes32, bool));
```
[the remark this answers is not in the document]
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:36]: you're okay with more gas?

File: lib/ERC6160
Comment (seunlanlege) [inline lib/ERC6160:1]: lets use the one in `evm/lib`

File: evm/src/modules/TokenGateway.sol (hunk 9 above)
Comment (AmadiMichael) [reply evm/src/modules/TokenGateway.sol:171]: yeah?

File: evm/src/GovernableToken.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (seunlanlege) [reply evm/src/GovernableToken.sol:36]: For governance calls i think it's fine

File: evm/script/DeployIsmp.s.sol (hunk 1 above)
Comment (AmadiMichael) [reply evm/script/DeployIsmp.s.sol:85]: unreachable code

File: evm/src/modules/HostManager.sol (hunk 2 above)
Comment (AmadiMichael) [reply evm/src/modules/HostManager.sol:59]: more gas?

File: evm/src/modules/TokenGateway.sol (hunk 3 above)
Comment (AmadiMichael) [reply evm/src/modules/TokenGateway.sol:93]: useful for batch actions, saves both runtime and deployment gas costs

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:34]: yeah?

File: evm/script/DeployIsmp.s.sol (hunk 1 above)
Comment (seunlanlege) [reply evm/script/DeployIsmp.s.sol:85]: actually it shouldn't be unreachable,

File: evm/src/modules/TokenGateway.sol (hunk 3 above)
Comment (seunlanlege) [reply evm/src/modules/TokenGateway.sol:93]: erc6160s can exist without an erc20 counterpart, lets split them out.

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (seunlanlege) [reply evm/src/GovernableToken.sol:34]: https://github.com/polytope-labs/hyperbridge/blob/main/evm/src/modules/TokenGateway.sol#L29-L34

File: evm/src/modules/TokenGateway.sol (hunk 3 above)
Comment (AmadiMichael) [reply evm/src/modules/TokenGateway.sol]: setting it to address(0) still maintains it, but okay, done

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:34]: this works for it too.

Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:34]: this uses roles to enforce it

File: evm/src/modules/HostManager.sol (hunk 2 above)
Comment (seunlanlege) [reply evm/src/modules/HostManager.sol:59]: yes, more safety

File: evm/src/GovernanceToken.sol (hunk 4 above)
Comment (seunlanlege) [reply evm/src/GovernanceToken.sol]: Why was this resolved, it's wrong

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (seunlanlege) [reply evm/src/GovernableToken.sol:34]: We only need to check that the caller is the host, not do a storage read

File: evm/src/GovernableToken.sol [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        _;
+    }
+
+    function _grantRole(bytes32 role, address account) private onlyRole(DEFAULT_ADMIN_ROLE) {
+        _roles[role][account] = true;
+    }
+
+    function _revokeRole(bytes32 role, address account) private onlyRole(DEFAULT_ADMIN_ROLE) {
+        _roles[role][account] = false;
+    }
```
Comment (seunlanlege) [inline evm/src/GovernableToken.sol:32]: delegate to the ERC6160 implemenatation

File: evm/src/modules/TokenGateway.sol [hunk 12]
```diff
@@ -41,58 +77,140 @@ contract TokenGateway is IIsmpModule {
...
+    function setTokenIdentifiersERC6160(bytes32 _tokenId, address _erc6160) external onlyAdmin {
+        _erc6160s[_tokenId] = _erc6160;
+    }
+
+    function setForeignTokenIdToLocalTokenId(bytes32 _foreignTokenId, bytes32 _localTokenId) external onlyAdmin {
+        _assets[_foreignTokenId] = _localTokenId;
+    }
+
+    function setChainsGateway(bytes memory _chain, address _host) external onlyAdmin {
+        _chainToGateway[_chain] = abi.encodePacked(_host);
```
Comment (seunlanlege) [inline evm/src/modules/TokenGateway.sol:103]: Can only be set once by admin. But can be updated through cross-chain governance

File: evm/src/modules/HostManager.sol (hunk 2 above)
Comment (AmadiMichael) [reply evm/src/modules/HostManager.sol:59]: it's already safe, the 1 byte (uint8) identifier of action to take is casted into an enum type, if the number is > than the max variant id of the enum, it reverts, so else means OnAcceptActions.SetHostParam is the only possible outcome if action != to the first one

File: evm/src/GovernanceToken.sol (hunk 4 above)
Comment (AmadiMichael) [reply evm/src/GovernanceToken.sol]: your comment is pointing at an old commit, check the commit after the commit the comment's is referencing

File: evm/src/GovernableToken.sol (hunk 6 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol]: this is also done and resolved

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:34]: the host is also in storage, so same thing.

File: evm/src/GovernableToken.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol]: this is also done and resolved already

File: evm/src/GovernableToken.sol (hunk 11 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol:32]: those are external functions, cannot be read by inheriting contracts.

File: evm/src/GovernableToken.sol (hunk 7 above)
Comment (seunlanlege) [reply evm/src/GovernableToken.sol:34]: prefer if we keep things consistent

Comment (AmadiMichael) [reply evm/src/GovernableToken.sol]: removed, it now uses the checks in the erc6160 called function

File: evm/src/GovernableToken.sol (hunk 11 above)
Comment (AmadiMichael) [reply evm/src/GovernableToken.sol]: removed

**Pull request merged**